packages/vscode-messenger-webview/src/messenger.ts

```diff
@@ -125,15 +162,28 @@ export class Messenger implements MessengerAPI {
         }
     }
 
-    sendRequest<P, R>(type: RequestType<P, R>, receiver: MessageParticipant, params?: P): Promise<R> {
+    sendRequest<P, R>(type: RequestType<P, R>, receiver: MessageParticipant, params?: P, cancelable?: CancellationTokenImpl): Promise<R> {
         if (receiver.type === 'broadcast') {
             throw new Error('Only notification messages are allowed for broadcast.');
         }
 
         const msgId = this.createMsgId();
-        const result = new Promise<R>((resolve, reject) => {
-            this.requests.set(msgId, { resolve: resolve as (value: unknown) => void, reject });
-        });
+        const pending = new DeferredRequest<R>();
+        this.requests.set(msgId, pending);
+        if (cancelable) {
+            cancelable.addCancelListener((reason) => {
+                // Send cancel message for pending request
+                this.vscode.postMessage(createCancelRequestMessage(receiver, msgId));
+                pending.reject(new Error(reason));
+                this.requests.delete(msgId);
+            });
+            pending.result.finally(() => {
+                // Request finished, nothing to do on cancel.
+                cancelable.clearCancelListeners();
```

Comment:
We shouldn't clear all listeners because the caller of this method might use the same cancellation token for multiple operations. Instead, we should remove only the listener that we added in this method.

I propose to do it in the same way as vscode does it: `addCancelListener` (or `onCancellationRequested` to be consistent with vscode) returns a `Disposable` which takes care of clearing the new listener.

Reply:
In fact `cancelable.clearCancelListeners` just removes the listeners from the `listeners` property inside the `CancellationTokenImpl` so they can be garbage collected.

Reply:
Now using Disposable to remove listeners.